### PR TITLE
Fix dynamic manifest processing

### PR DESCRIPTION
**Pull request by randyrossi**
This CL fixes issue 1: Dynamic manifests (LIVE) causes SEGFAULT
after 1st refresh (https://github.com/google/ndash/issues/1).

NDash was crashing on YouTube manifests if /mpd_version/4 or
mpd_version=4 was omited from the manifest url. This tells youtube to
revert back to v1 behavior which is essentially broken and produces bad
manifests.

In YouTube's v1 manifests, the period's start time increases after every
refresh which causes NDash to think it's a new period each time.  There is
no 'id' on a period so the only thing that uniquely identifies it is its
start time.  (Period's start times are not supposed to change and it
appears YouTube's initial implementation of a live window mistakenly
updates the start time on each refresh.)

This, in turn, caused NDash to remove the previous period from its
period holder list causing objects owned by the period/representation holders
to be destroyed while loaders were still using them. (Most notibly was
the ChunkExtractorWrapper which held the stream parser, media log, etc.)
The representation holder's unique ptr to ChunkExtractorWrapper was changed
to a scoped ref ptr so that other parts of the code can be sure they
will still be alive even if the representation holder happens to get
destroyed.  However, as long as incoming manifests are well behaved,
this should actually never happen.

For additional safety, I also gave the representation holders their own
copy of the media format instead of holding on to pointers from the
manifest which might get destroyed while the holder is still alive.

This CL doesn't really address the underlying issue which is that
pointers to objects owned by the manifest might get passed around to
other parts of the code that expect them to live beyond the life of the
manifest.  Likewise for Period/Representation holders.  This CL plugs the
known holes but others may crop up in the future.

Also discovered a line of code that was missing from the original port
in representation holder where the index should have been updated.

### CODE REVIEW

File: ndash/src/chunk/container_media_chunk.cc [hunk 1]
```diff
@@ -198,12 +200,12 @@ void ContainerMediaChunk::GiveSeekMap(
...
-std::unique_ptr<MediaFormat> ContainerMediaChunk::GetAdjustedMediaFormat(
-    const MediaFormat* format,
+std::unique_ptr<const MediaFormat> ContainerMediaChunk::GetAdjustedMediaFormat(
+    std::unique_ptr<const MediaFormat> format,
     base::TimeDelta sample_offset) {
-  std::unique_ptr<MediaFormat> out;
+  std::unique_ptr<const MediaFormat> out;
 
-  if (format == nullptr) {
+  if (format.get() == nullptr) {
```
Comment (rdaum) [inline ndash/src/chunk/container_media_chunk.cc]: nit: You should be able to just do if (!format) ...

File: ndash/src/chunk/container_media_chunk_unittest.cc [hunk 2]
```diff
@@ -228,20 +231,21 @@ TEST(ContainerMediaChunkTest, TestLoadFail) {
   constexpr int64_t kTestEndTime = 2;
   constexpr int32_t kTestChunkIndex = 3;
   const base::TimeDelta kSampleOffset = base::TimeDelta::FromMicroseconds(5000);
-  const std::unique_ptr<MediaFormat> media_format =
-      MediaFormat::CreateVideoFormat("1", "video/mp4", "h264", 2200000, 32768,
-                                     1234567, 640, 480, nullptr, 16, 45, 1.666);
+  std::unique_ptr<MediaFormat> media_format(MediaFormat::CreateVideoFormat(
+      "1", "video/mp4", "h264", 2200000, 32768, 1234567, 640, 480, nullptr, 16,
```
Comment (rdaum) [inline ndash/src/chunk/container_media_chunk_unittest.cc:235]: q: are these two constants ("video/mp4" and "h264") present in chrome's media lib somewhere?

File: ndash/src/dash/representation_holder.h [hunk 3]
```diff
@@ -45,32 +46,38 @@ class RepresentationHolder {
...
+  const MediaFormat* media_format() const { return media_format_.get(); }
 
   void GiveMediaFormat(std::unique_ptr<const MediaFormat> media_format);
   void GiveSegmentIndex(
       std::unique_ptr<const mpd::DashSegmentIndexInterface> segment_index);
 
   // Called when the Manifest is updated and the representation pointer needs
   // to be updated.
-  // Returns true if successful, false otherwise (BehindLiveWindowException)
+  // Returns true if succlogcessful, false otherwise (BehindLiveWindowException)
```
Comment (rdaum) [inline ndash/src/dash/representation_holder.h]: typo here

File: ndash/src/chunk/container_media_chunk_unittest.cc (hunk 2 above)
Comment (randyrossi) [reply ndash/src/chunk/container_media_chunk_unittest.cc:235]: No. There aren't.

File: ndash/src/dash/representation_holder.h (hunk 3 above)
Comment (randyrossi) [reply ndash/src/dash/representation_holder.h]: Done

File: ndash/src/chunk/container_media_chunk.cc (hunk 1 above)
Comment (randyrossi) [reply ndash/src/chunk/container_media_chunk.cc]: Done

File: ndash/src/chunk/container_media_chunk_unittest.cc (hunk 2 above)
Comment (rdaum) [reply ndash/src/chunk/container_media_chunk_unittest.cc:235]: Ack.

**Pull request merged**